File: src/jev_factorio/lua/fair_actions.lua
local function configured_player_index()
    local index = storage.jev_player_index
    if index == nil then index = 1 end
    assert(type(index) == "number" and index > 0 and index % 1 == 0,
        "Native player index must be a positive integer")
    return index
end

-- agent_characters[1] remains FLE's logical agent slot. The native player
-- owning that character may have a different index, fixed for this runtime.
local player_index = configured_player_index()
assert(storage.jev_bound_player_index == nil or storage.jev_bound_player_index == player_index,
    "Native player index cannot change within a runtime")
storage.jev_bound_player_index = player_index
local function selected_player()
    assert(configured_player_index() == player_index
        and storage.jev_bound_player_index == player_index,
        "Native player index cannot change within a runtime")
    return game.get_player(player_index)
end

storage.fair = storage.fair or {}
local fair = storage.fair
fair.quarantined = true
script.on_nth_tick(5, nil)
script.on_nth_tick(15, nil)
script.on_nth_tick(60, nil)

if storage.actions and storage.actions.inspect_inventory
    and storage.actions.inspect_inventory ~= fair.inspect_inventory then
    local inspect_inventory = storage.actions.inspect_inventory
    fair.inspect_inventory = function(...)
        local result = table.pack(pcall(inspect_inventory, ...))
        script.on_nth_tick(60, nil)
        if not result[1] then error(result[2]) end
        return table.unpack(result, 2, result.n)
    end
    storage.actions.inspect_inventory = fair.inspect_inventory
end

fair.actor = function()
    local player = selected_player()
    local character = storage.agent_characters and storage.agent_characters[1]
    assert(player and player.connected and character and character.valid,
        "Fair play requires the original connected character")
    assert(player.character == character, "Fair player binding changed")
    assert(game.speed == 1 and not player.cheat_mode, "Fair play requires normal game speed")
    return player
end

fair.stop = function(reason)
    -- Cleanup must still stop the original player when configuration drifts.
    local player = game.get_player(player_index)
    if player and player.character == (storage.agent_characters or {})[1] then
        player.walking_state = {walking = false}
        player.mining_state = {mining = false}
    end
    if fair.job and (fair.job.status ~= "failed" or reason) then
        fair.job.status = reason and "failed" or "completed"
        fair.job.error = reason
    end
end

fair.bind = function()
    fair.quarantined = true
    script.on_nth_tick(5, nil)
    script.on_nth_tick(15, nil)
    script.on_nth_tick(60, nil)
    for _, name in pairs({"crafting_queue", "harvest_queues", "walking_queues"}) do
        assert(not storage[name] or next(storage[name]) == nil,
            "Legacy scripted work must be reconciled: " .. name)
    end
    local player = selected_player()
    local character = (storage.agent_characters or {})[1]
    assert(player and player.connected and character and character.valid,
        "Fair play requires the original connected character")
    assert(not player.character or player.character == character,
        "Refusing to replace a different character")
    if not player.character then
        player.set_controller{type = defines.controllers.character, character = character}
    end
    local player = fair.actor()
    storage.fast = false
    fair.stop("Controls stopped on adapter attachment")
    fair.quarantined = false
    return {position = player.position}
end

-- Rejected exact start/goal routes are cooled down, not retried every tick.
-- No assumed API flag can prohibit every neutral destructible obstacle: inspect
-- every PathfinderWaypoint before assigning any walking controls.
fair.blocked_routes = fair.blocked_routes or {}
local function route_key(player, position)
    return table.concat({player.surface.index or 0, player.character.unit_number,
        player.position.x, player.position.y, position.x, position.y}, ":")
end
local function reject_route(job, code, reason)
    job.failure_code = code
    job.movement_started = job.movement_started or false
    fair.blocked_routes[job.route_key] = game.tick + 3600
    local count = 0
    for key, expiry in pairs(fair.blocked_routes) do
        if expiry <= game.tick then fair.blocked_routes[key] = nil else count = count + 1 end
    end
    -- Keep memory bounded even across many distinct destinations.
    while count > 64 do
        local oldest_key, oldest_expiry
        for key, expiry in pairs(fair.blocked_routes) do
            if not oldest_expiry or expiry < oldest_expiry then
                oldest_key, oldest_expiry = key, expiry
            end
        end
        fair.blocked_routes[oldest_key] = nil
        count = count - 1
    end
    fair.stop(reason)
end

-- Plan both legs of at most four detours BEFORE moving. Nine native requests
-- (one original plus two per detour) and a 600-tick planning deadline bound work.
local function request_move_path(job, player, start, goal)
    assert((job.path_requests or 0) < 9, "Native path request budget exhausted")
    job.path_requests = (job.path_requests or 0) + 1
    job.request = player.surface.request_path{
        bounding_box = player.character.prototype.collision_box,
        collision_mask = player.character.prototype.collision_mask,
        start = start, goal = goal, force = player.force,
        radius = 0.2, entity_to_ignore = player.character, can_open_gates = true,
        pathfind_flags = {cache = false, allow_paths_through_own_entities = false,
                          allow_destroy_friendly_entities = false}
    }
end
local function next_detour(job, code)
    local player = fair.actor()
    local offsets = {6, -6, 12, -12}
    job.detour_index = (job.detour_index or 0) + 1
    if job.detour_index > #offsets or job.path_requests >= 9
        or game.tick > job.path_deadline then
        reject_route(job, code,
            "No safe route within bounded detours; choose another target or clear obstacles by normal mining")
        return
    end
    local dx, dy = job.goal.x - job.origin.x, job.goal.y - job.origin.y
    local length = math.sqrt(dx * dx + dy * dy)
    if length == 0 then
        reject_route(job, code, "No safe zero-length route; inspect the target obstruction")
        return
    end
    local center = job.blocked_waypoint or {
        x = (job.origin.x + job.goal.x) / 2,
        y = (job.origin.y + job.goal.y) / 2
    }
    job.detour_goal = {x = center.x - dy / length * offsets[job.detour_index],
                       y = center.y + dx / length * offsets[job.detour_index]}
    job.detour_prefix, job.path_leg = nil, 1
    request_move_path(job, player, job.origin, job.detour_goal)
end

fair.begin_move = function(position)
    local player = fair.actor()
    fair.stop()
    local character = player.character
    fair.job = {
        kind = "walk", status = "path_pending", lease = game.tick + 180,
        unit = character.unit_number, last_progress = game.tick,
        last_position = player.position, goal = position,
        route_key = route_key(player, position), movement_started = false,
        origin = {x = player.position.x, y = player.position.y},
        path_requests = 0, path_deadline = game.tick + 600
    }
    if (fair.blocked_routes[fair.job.route_key] or 0) > game.tick then
        fair.job.failure_code = "blocked_route_cooldown"
        fair.stop("Known blocked route is cooling down; choose another target or clear it normally")
        return {rejected = true, failure_code = fair.job.failure_code, movement_started = false}
    end
    request_move_path(fair.job, player, fair.job.origin, position)
    return {request = fair.job.request}
end

local function mining_entity(player, position, item)
    local filter = {position = position, radius = 0.75}
    if item == "wood" then filter.type = "tree" else filter.name = item end
    local best, best_distance
    for _, entity in pairs(player.surface.find_entities_filtered(filter)) do
        if entity.valid and entity.minable then
            local distance = (entity.position.x - position.x)^2
                + (entity.position.y - position.y)^2
            if not best or distance < best_distance then
                best, best_distance = entity, distance
            end
        end
    end
    assert(best, "No mineable resource at target")
    return best
end

fair.mine_approach = function(position, item)
    local player = fair.actor()
    local entity = mining_entity(player, position, item)
    if player.can_reach_entity(entity) then return {reachable = true} end
    local horizontal = player.position.x - entity.position.x
    local vertical = player.position.y - entity.position.y
    local distance = math.sqrt(horizontal * horizontal + vertical * vertical)
    assert(distance > 0, "Unreachable mining target overlaps player")
    local target = {
        x = entity.position.x + horizontal / distance * 1.5,
        y = entity.position.y + vertical / distance * 1.5
    }
    local approach = player.surface.find_non_colliding_position("character", target, 2, 0.25)
    assert(approach, "No collision-free mining approach")
    return {reachable = false, position = approach}
end

fair.begin_mine = function(position, item, quantity)
    local player = fair.actor()
    fair.stop()
    local entity = mining_entity(player, position, item)
    assert(player.can_reach_entity(entity), "Mining target is outside normal reach")
    player.update_selected_entity(entity.position)
    assert(player.selected == entity, "Mining target is obscured by another entity")
    fair.job = {
        kind = "mine", status = "mining", lease = game.tick + 180,
        unit = player.character.unit_number, entity = entity, item = item,
        baseline = player.get_item_count(item), quantity = quantity,
        last_progress = game.tick, last_count = player.get_item_count(item)
    }
    player.mining_state = {mining = true, position = entity.position}
    return {baseline = fair.job.baseline}
end

fair.next_mine_target = function(item, radius)
    local player = fair.actor()
    assert(type(item) == "string", "Mining item must be a string")
    assert(type(radius) == "number" and radius > 0 and radius <= 128,
        "Mining search radius is invalid")
    local filter = {position = player.position, radius = radius}
    if item == "wood" then filter.type = "tree" else filter.name = item end
    local best, best_distance
    for _, entity in pairs(player.surface.find_entities_filtered(filter)) do
        if entity.valid and entity.minable then
            local horizontal = entity.position.x - player.position.x
            local vertical = entity.position.y - player.position.y
            local distance = horizontal * horizontal + vertical * vertical
            if not best or distance < best_distance then
                -- Reject targets obscured by another entity before committing
                -- an observation.  This only updates the normal cursor; it
                -- does not walk, mine, transfer, or alter game speed.  The
                -- later begin_mine call independently enforces normal reach.
                player.update_selected_entity(entity.position)
                if player.selected == entity then
                    best, best_distance = entity, distance
                end
            end
        end
    end
    if not best then return {} end
    return {
        position = {x = best.position.x, y = best.position.y},
        unit_number = best.unit_number,
        name = best.name,
        surface_index = best.surface.index,
    }
end

fair.discover_mine_target = function(item, center, radius)
    -- Discovery inspects only terrain the campaign already generated. It may
    -- probe normal cursor selection, but never starts walking or mining. The
    -- later fair harvesting path walks to the returned entity
    -- and independently verifies normal reach and cursor selection.
    local player = fair.actor()
    assert(type(item) == "string", "Mining item must be a string")
    assert(type(center) == "table" and type(center.x) == "number"
        and type(center.y) == "number", "Mining search center is invalid")
    assert(type(radius) == "number" and radius > 0 and radius <= 1024,
        "Mining discovery radius is invalid")
    local filter = {position = center, radius = radius}
    if item == "wood" then filter.type = "tree" else filter.name = item end
    local best, best_distance
    for _, entity in pairs(player.surface.find_entities_filtered(filter)) do
        if entity.valid and entity.minable
            and (entity.type ~= "resource" or entity.amount > 0) then
            local horizontal = entity.position.x - center.x
            local vertical = entity.position.y - center.y
            local distance = horizontal * horizontal + vertical * vertical
            if not best or distance < best_distance then
                player.update_selected_entity(entity.position)
                if player.selected == entity then
                    best, best_distance = entity, distance
                end
            end
        end
    end
    if not best then return {} end
    return {
        position = {x = best.position.x, y = best.position.y},
        unit_number = best.unit_number,
        name = best.name,
        surface_index = best.surface.index,
    }
end

fair.observe = function()
    local player = fair.actor()
    local job = fair.job or {}
    if job.status ~= "failed" and job.status ~= "completed" then
        job.lease = game.tick + 180
    end
    return {
        position = player.position, tick = game.tick, status = job.status or "idle",
        error = job.error, failure_code = job.failure_code,
        movement_started = job.movement_started or false, path_requests = job.path_requests or 0,
        walking = player.walking_state.walking,
        mining = player.mining_state.mining,
        gained = job.item and player.get_item_count(job.item) - job.baseline or 0
    }
end

fair.find_build_site = function(name, center, radius)
    local player = fair.actor()
    assert(type(name) == "string" and prototypes.entity[name], "Unknown building prototype")
    assert(type(center) == "table" and type(center.x) == "number"
        and type(center.y) == "number", "Invalid build-site center")
    assert(type(radius) == "number" and radius >= 0 and radius <= 32
        and radius % 0.5 == 0, "Invalid build-site radius")
    local directions = {
        defines.direction.north, defines.direction.east,
        defines.direction.south, defines.direction.west
    }
    local best, best_distance
    local half_steps = radius * 2
    for horizontal = -half_steps, half_steps do
        for vertical = -half_steps, half_steps do
            local position = {
                x = center.x + horizontal / 2,
                y = center.y + vertical / 2
            }
            local distance = horizontal * horizontal + vertical * vertical
            for _, direction in ipairs(directions) do
                if player.surface.can_place_entity{
                    name = name, position = position, direction = direction,
                    force = player.force,
                    build_check_type = defines.build_check_type.manual
                } and (not storage.campaign or not storage.campaign.production_reserved
                    or not storage.campaign.production_reserved(name, position, direction))
                    and (not storage.campaign or not storage.campaign.mining_outpost_reserved
                    or not storage.campaign.mining_outpost_reserved(name, position, direction))
                    and (not best or distance < best_distance) then
                    best = {position = position, direction = direction}
                    best_distance = distance
                end
            end
        end
    end
    assert(best, "No ordinary build site")
    return best
end

fair.place = function(name, position, direction)
    local player = fair.actor()
    assert(not player.surface.find_entity(name, position), "Building already exists")
    assert((player.position.x - position.x)^2 + (player.position.y - position.y)^2
        <= player.build_distance^2, "Building is outside normal reach")
    assert(player.clear_cursor(), "Cannot clear the cursor without losing items")
    local before = player.get_item_count(name)
    local stack = player.get_main_inventory().find_item_stack(name)
    assert(stack and stack.valid_for_read, "Missing building item")
    assert(player.cursor_stack.transfer_stack(stack), "Cannot move existing item into cursor")
    local ok, failure = pcall(function()
        assert(player.can_build_from_cursor{position = position, direction = direction},
            "Building is obstructed or outside normal reach")
        player.build_from_cursor{position = position, direction = direction}
    end)
    local cleared = player.clear_cursor()
    assert(ok, failure)
    assert(cleared, "Could not return cursor items")
    assert(player.get_item_count(name) == before - 1, "Native build did not consume one item")
    local entity = player.surface.find_entity(name, position)
    assert(entity and entity.valid and entity.force == player.force, "Native build did not create entity")
    return {name = entity.name, position = entity.position, unit_number = entity.unit_number,
        drop_position = entity.type == "mining-drill" and entity.drop_position or nil}
end

fair.insert = function(name, position, item, quantity)
    local player = fair.actor()
    local entity = player.surface.find_entity(name, position)
    assert(entity and entity.valid and player.can_reach_entity(entity),
        "Interaction target is outside normal reach")
    local inventory = player.get_main_inventory()
    assert(inventory.get_item_count(item) >= quantity, "Missing transfer items")
    assert(entity.can_insert{name = item, count = quantity}, "Transfer destination is full")
    local removed = inventory.remove{name = item, count = quantity}
    local inserted = entity.insert{name = item, count = removed}
    if inserted < removed then
        assert(inventory.insert{name = item, count = removed - inserted} == removed - inserted)
    end
    assert(inserted == quantity, "Partial transfer requires reconciliation")
    return {quantity = inserted}
end

local previous_path = script.get_event_handler(defines.events.on_script_path_request_finished)
if previous_path ~= fair.path_handler then fair.previous_path = previous_path end
fair.path_handler = function(event)
    local job = fair.job
    if job and job.status == "path_pending" and event.id == job.request then
        if not event.path or #event.path == 0 then
            next_detour(job, event.try_again_later and "pathfinder_busy" or "no_safe_path")
            return
        end
        for _, waypoint in ipairs(event.path) do
            if waypoint.needs_destroy_to_reach then
                job.blocked_waypoint = job.blocked_waypoint or waypoint.position
                next_detour(job, "destruction_required")
                return
            end
        end
        if job.path_leg == 1 then
            job.detour_prefix, job.path_leg = event.path, 2
            request_move_path(job, fair.actor(), event.path[#event.path].position, job.goal)
            return
        end
        local path = event.path
        if job.path_leg == 2 then
            path = job.detour_prefix
            for _, point in ipairs(event.path) do path[#path + 1] = point end
        end
        local player = fair.actor()
        if (player.position.x - job.origin.x)^2 + (player.position.y - job.origin.y)^2 > 0.0625 then
            reject_route(job, "route_origin_changed", "Actor moved during route planning; reconcile before retry")
            return
        end
        job.last_progress, job.last_position = game.tick, player.position
        job.path, job.index, job.status = path, 1, "walking"
    elseif fair.previous_path then fair.previous_path(event) end
end
script.on_event(defines.events.on_script_path_request_finished, fair.path_handler)

local previous_tick = script.get_event_handler(defines.events.on_tick)
if previous_tick ~= fair.tick_handler then fair.previous_tick = previous_tick end
fair.tick_handler = function(event)
    if fair.quarantined then fair.stop("Adapter attachment is not validated"); return end
    local job = fair.job
    if not job or job.status == "failed" or job.status == "completed" then return end
    local ok, player = pcall(fair.actor)
    if not ok then fair.stop("Fair player/session invariant failed"); return end
    if player.character.unit_number ~= job.unit or game.tick > job.lease then
        fair.stop("Control lease expired or character changed"); return
    end
    if job.status == "path_pending" and game.tick > job.path_deadline then
        reject_route(job, "path_deadline", "Native path planning exceeded its bounded deadline")
        return
    end
    if job.status == "mining" then
        local count = player.get_item_count(job.item)
        if count - job.baseline >= job.quantity then fair.stop(); return end
        if not job.entity.valid then fair.stop("Resource depleted before requested amount"); return end
        if not player.can_reach_entity(job.entity) then fair.stop("Mining target left reach"); return end
        player.update_selected_entity(job.entity.position)
        if player.selected ~= job.entity then fair.stop("Mining target became obscured"); return end
        if count ~= job.last_count then job.last_count, job.last_progress = count, game.tick end
        if game.tick - job.last_progress > 600 then fair.stop("Native mining made no progress"); return end
        player.mining_state = {mining = true, position = job.entity.position}
    elseif job.status == "walking" then
        local point = job.path[job.index]
        while point and (player.position.x - point.position.x)^2
            + (player.position.y - point.position.y)^2 < 0.0625 do
            job.index = job.index + 1
            point = job.path[job.index]
        end
        if not point then fair.stop(); return end
        local horizontal = point.position.x - player.position.x
        local vertical = point.position.y - player.position.y
        local direction
        if math.abs(horizontal) > 2 * math.abs(vertical) then
            direction = horizontal > 0 and defines.direction.east or defines.direction.west
        elseif math.abs(vertical) > 2 * math.abs(horizontal) then
            direction = vertical > 0 and defines.direction.south or defines.direction.north
        elseif horizontal > 0 then
            direction = vertical > 0 and defines.direction.southeast or defines.direction.northeast
        else
            direction = vertical > 0 and defines.direction.southwest or defines.direction.northwest
        end
        if (player.position.x - job.last_position.x)^2
            + (player.position.y - job.last_position.y)^2 > 0.25 then
            job.last_progress, job.last_position = game.tick, player.position
        end
        if game.tick - job.last_progress > 300 then
            reject_route(job, "movement_obstructed", "Native walking is obstructed; partial movement must be reconciled")
            return
        end
        job.movement_started = true
        player.walking_state = {walking = true, direction = direction}
    end
end
script.on_event(defines.events.on_tick, fair.tick_handler)

File: src/jev_factorio/backends/fair_actions.py
"""Player-controlled actions without teleporting or synthetic harvests."""
from __future__ import annotations

import json
import math
import time
from importlib.resources import files
from types import SimpleNamespace
from typing import Any
from .errors import ConnectionPreflightRejected


class FairActions:
    def __init__(self, backend: Any) -> None:
        self.backend = backend
        self.command(files("jev_factorio").joinpath("lua/fair_actions.lua").read_text())
        self.call("bind")

    def command(self, script: str) -> str:
        result = self.backend._instance.rcon_client.send_command("/sc " + script) or ""
        if result.startswith("Cannot execute command."):
            raise RuntimeError(result)
        return result

    def call(self, function: str, *arguments: Any) -> dict:
        encoded = ", ".join(
            "helpers.json_to_table(" + json.dumps(json.dumps(value, allow_nan=False)) + ")"
            for value in arguments
        )
        return json.loads(self.command(
            f"rcon.print(helpers.table_to_json(storage.fair.{function}({encoded})))"
        ))

    @staticmethod
    def position(position: Any) -> dict:
        result = {"x": float(position.x), "y": float(position.y)}
        if not all(math.isfinite(value) for value in result.values()):
            raise ValueError("Position must be finite")
        return result

    def wait(self, timeout: float = 180) -> dict:
        deadline = time.monotonic() + timeout
        try:
            while time.monotonic() < deadline:
                state = self.call("observe")
                if state["status"] == "completed":
                    return state
                if state["status"] == "failed":
                    raise RuntimeError(state.get("error", "Native controls failed"))
                time.sleep(0.1)
            raise TimeoutError("Native action exceeded its bounded observation window")
        finally:
            self.command("storage.fair.stop()")

    def _note(self, key: str, amount: int = 1) -> None:
        if not hasattr(self, "metrics"):
            self.metrics = {}
        self.metrics[key] = self.metrics.get(key, 0) + amount

    def move_to(self, position: Any) -> Any:
        from fle.env import Position

        self._note("move_requests")
        self.call("begin_move", self.position(position))
        state = self.wait()
        result = Position(**state["position"])
        self.backend._instance.namespace.player_location = result
        return result

    def approach(self, position: Any, name: str = "character") -> None:
        from fle.env import Position

        center = self.position(position)
        self._note("approach_requests")
        result = json.loads(self.command(
            "local player = storage.fair.actor(); local prototype = prototypes.entity["
            + json.dumps(name) + "]; local box = prototype.selection_box; "
            "local target = helpers.json_to_table(" + json.dumps(json.dumps(center)) + "); "
            "local entity = player.surface.find_entity(" + json.dumps(name) + ", target); "
            "if entity and entity.valid and player.can_reach_entity(entity) then "
            "rcon.print(helpers.table_to_json({reachable=true})); return end; "
            "target.x = target.x + math.max(math.abs(box.left_top.x), "
            "math.abs(box.right_bottom.x)) + 1.5; "
            "local position = player.surface.find_non_colliding_position('character', target, 8, 0.25); "
            "assert(position, 'No collision-free approach'); rcon.print(helpers.table_to_json(position))"
        ))
        if result.get("reachable") is True:
            self._note("approaches_skipped_in_reach")
            return
        self.move_to(Position(**result))

    def harvest(self, resource: str, position: Any, quantity: int) -> int:
        from fle.env import Position

        if type(quantity) is not int or quantity <= 0:
            raise ValueError("Mining quantity must be positive")
        self._note("mining_batches_started")
        gained = 0
        for attempt in range(quantity):
            if gained == 0:
                # Honor the resource coordinate from the observation that
                # authorized this action.  It is the target to which the
                # controller committed, rather than merely a same-name node
                # near the player.
                target = self.position(position)
            else:
                # After native mining depleted a node, reacquire around the
                # actor rather than searching around the stale original
                # coordinate. This only selects a target: approach() still
                # walks normally and begin_mine() enforces reach.
                target = self.call("next_mine_target", resource, 64).get("position")
                if not isinstance(target, dict):
                    raise RuntimeError("No mineable resource observed near the walking actor")
            approach = self.call("mine_approach", target, resource)
            if not approach["reachable"]:
                self.move_to(Position(**approach["position"]))
            self._note("mining_starts")
            self.call("begin_mine", target, resource, quantity - gained)
            try:
                result = self.wait()
            except RuntimeError:
                result = self.call("observe")
                if result.get("error") != "Resource depleted before requested amount" or result["gained"] <= 0:
                    raise
            gained += result["gained"]
            self._note("mined_items", result["gained"])
            if gained >= quantity:
                return gained
        raise RuntimeError("Mining target budget exhausted")

    def approach_build(self, position: Any) -> None:
        """Walk only when the native placement center is outside build reach."""
        from fle.env import Position

        center = self.position(position)
        self._note("approach_requests")
        result = json.loads(self.command(
            "local player=storage.fair.actor(); local target=helpers.json_to_table("
            + json.dumps(json.dumps(center)) + "); "
            "local dx=player.position.x-target.x; local dy=player.position.y-target.y; "
            "local distance_squared=dx*dx+dy*dy; local reach=player.build_distance; "
            "if distance_squared<=reach^2 then "
            "rcon.print(helpers.table_to_json({reachable=true})); return end; "
            "assert(reach>0, 'No native build reach'); "
            "local distance=math.sqrt(distance_squared); local radius=math.max(0,reach-0.5); "
            "local near={x=target.x+dx/distance*radius,y=target.y+dy/distance*radius}; "
            "local position=player.surface.find_non_colliding_position('character',near,1,0.25); "
            "assert(position, 'No collision-free build approach'); "
            "assert((position.x-target.x)^2+(position.y-target.y)^2<=reach^2, "
            "'Build approach is outside normal reach'); "
            "rcon.print(helpers.table_to_json(position))"
        ))
        if result.get("reachable") is True:
            self._note("approaches_skipped_in_reach")
            return
        self.move_to(Position(**result))

    def place_entity(self, prototype: Any, position: Any, direction: Any,
                     exact: bool = False) -> Any:
        from fle.env import Position

        name = prototype.value[0]
        target = self.position(position)
        direction_value = direction.value
        if not exact:
            site = self.call("find_build_site", name, target, 8)
            target, direction_value = site["position"], site["direction"]
        self.approach_build(Position(**target))
        result = self.call("place", name, target, direction_value)
        return SimpleNamespace(
            name=result["name"], position=Position(**result["position"]),
            drop_position=Position(**result["drop_position"]) if result.get("drop_position") else None,
        )

    def insert_item(self, prototype: Any, entity: Any, quantity: int) -> int:
        self.approach(entity.position, entity.name)
        return self.call("insert", entity.name, self.position(entity.position),
                         prototype.value[0], quantity)["quantity"]

    @staticmethod
    def _connection_corridor(start: dict, end: dict, *, horizontal_first: bool) -> list[tuple[int, int, int, int]]:
        """Bound native placement discovery to one collision-aware L corridor.

        Scanning the complete rectangle between two distant endpoints makes a
        narrow, ordinary pole route depend on its bounding-box area.  The
        corridor keeps the search bounded while retaining eight cells of local
        detour room around each leg.  A caller can try the opposite L without
        widening either search into a remote-placement shortcut.
        """
        source_x, source_y = math.floor(start["x"]), math.floor(start["y"])
        target_x, target_y = math.floor(end["x"]), math.floor(end["y"])
        left, right = min(source_x, target_x) - 8, max(source_x, target_x) + 8
        top, bottom = min(source_y, target_y) - 8, max(source_y, target_y) + 8
        if horizontal_first:
            return [
                (left, right, source_y - 8, source_y + 8),
                (target_x - 8, target_x + 8, top, bottom),
            ]
        return [
            (source_x - 8, source_x + 8, top, bottom),
            (left, right, target_y - 8, target_y + 8),
        ]

    def _connection_cells(self, name: str, fluid: str,
                          rectangles: list[tuple[int, int, int, int]]) -> tuple[set, set]:
        searched = sum((right - left + 1) * (bottom - top + 1)
                       for left, right, top, bottom in rectangles)
        if searched > 16_384:
            raise ValueError("Connection search exceeds bounded area")
        loops = "".join(
            f"for horizontal={left},{right} do for vertical={top},{bottom} do "
            "include(horizontal, vertical) end end; "
            for left, right, top, bottom in rectangles
        )
        cells = json.loads(self.command(
            "local player = storage.fair.actor(); local result = {buildable={}, existing={}}; "
            "local seen = {}; local function include(horizontal, vertical) "
            "local key = horizontal .. ':' .. vertical; if seen[key] then return end; seen[key] = true; "
            "local position = {x=horizontal+0.5,y=vertical+0.5}; "
            "local entity = player.surface.find_entity(" + json.dumps(name) + ", position); "
            "if entity and entity.force == player.force then "
            "local contents = #entity.fluidbox > 0 and entity.fluidbox[1]; "
            "if not contents or contents.name == " + json.dumps(fluid) + " then "
            "table.insert(result.existing, position) end "
            "elseif player.surface.can_place_entity{name=" + json.dumps(name)
            + ", position=position, force=player.force, build_check_type=defines.build_check_type.manual} "
            "then table.insert(result.buildable, position) end; end; "
            + loops + "rcon.print(helpers.table_to_json(result))"
        ))
        return (
            {(point["x"], point["y"]) for point in cells["buildable"]},
            {(point["x"], point["y"]) for point in cells["existing"]},
        )

    def connect(self, source: Any, target: Any, prototype: Any, fluid: str = "") -> None:
        from fle.env import Direction, Position
        from ..planning.connections import (
            select_pole_positions,
            shortest_pipe_path,
            shortest_wire_path,
        )

        name = prototype.value[0]
        if name not in {"pipe", "small-electric-pole"}:
            raise ValueError("Unsupported fair connection type")
        start = self.position(getattr(source, "position", source))
        end = self.position(getattr(target, "position", target))
        route = None
        route_error = None
        for horizontal_first in (True, False):
            buildable, existing = self._connection_cells(
                name, fluid, self._connection_corridor(
                    start, end, horizontal_first=horizontal_first,
                ),
            )
            origin, destination = (start["x"], start["y"]), (end["x"], end["y"])
            if name == "small-electric-pole":
                candidates = buildable | existing
                if not candidates:
                    route_error = ValueError("No ordinary pole placement cells")
                    continue
                origin = min(candidates, key=lambda point: math.dist(point, origin))
                destination = min(candidates, key=lambda point: math.dist(point, destination))
                if math.dist(origin, (start["x"], start["y"])) > 3.5 or math.dist(
                    destination, (end["x"], end["y"])
                ) > 3.5:
                    route_error = ValueError("No nearby ordinary pole placement")
                    continue
            try:
                if name == "small-electric-pole":
                    route = shortest_wire_path(
                        origin, destination, buildable, existing, max_wire_distance=6
                    )
                else:
                    route = shortest_pipe_path(origin, destination, buildable, existing)
                break
            except ValueError as error:
                route_error = error
        if route is None:
            raise ConnectionPreflightRejected("no_connection_route")
        if name == "small-electric-pole":
            route = select_pole_positions(route, max_wire_distance=6)
        required = sum(point not in existing for point in route)
        available = json.loads(self.command(
            "rcon.print(helpers.table_to_json({count=storage.fair.actor().get_item_count("
            + json.dumps(name) + ")}))"
        ))["count"]
        if available < required:
            raise ConnectionPreflightRejected("insufficient_connection_materials")
        for horizontal, vertical in route:
            if (horizontal, vertical) not in existing:
                self.place_entity(prototype, Position(x=horizontal, y=vertical),
                                  direction=Direction.UP, exact=True)
